libsignal-service/src/groups_v2/manager.rs

```diff
@@ -0,0 +1,231 @@
+use std::{
+    collections::HashMap,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use crate::{
+    configuration::Endpoint,
+    prelude::{PushService, ServiceError},
+    proto::DecryptedGroup,
+    push_service::HttpCredentials,
+};
+
+use rand::RngCore;
+use serde::Deserialize;
+use uuid::Uuid;
+use zkgroup::{
+    auth::AuthCredentialResponse, groups::GroupSecretParams, ServerPublicParams,
+};
+
+use super::operations::GroupOperations;
+
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TemporalCredential {
+    credential: String,
+    redemption_time: i64,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CredentialResponse {
+    credentials: Vec<TemporalCredential>,
+}
+
+impl CredentialResponse {
+    pub fn parse(
+        self,
+    ) -> Result<HashMap<i64, AuthCredentialResponse>, ServiceError> {
+        Ok(self
+            .credentials
+            .into_iter()
+            .map(|c| {
+                let bytes = base64::decode(c.credential)?;
+                let data = bincode::deserialize(&bytes)?;
+                Ok((c.redemption_time, data))
+            })
+            .collect::<Result<_, ServiceError>>()?)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum CredentialsCacheError {
+    #[error("failed to read values from cache: {0}")]
+    ReadError(String),
+    #[error("failed to write values from cache: {0}")]
+    WriteError(String),
+}
+
+/// Global cache for groups v2 credentials, as demonstrated in the libsignal-service
+/// java library of Signal-Android.
+///
+/// A basic in-memory implementation is provided with `InMemoryCredentialsCache`.
+pub trait CredentialsCache {
+    fn clear(&mut self) -> Result<(), CredentialsCacheError>;
+
+    /// Get an entry of the cache, key usually represents the day number since EPOCH.
+    fn get(
+        &self,
+        key: &i64,
+    ) -> Result<Option<&AuthCredentialResponse>, CredentialsCacheError>;
+
+    /// Overwrite the entire contents of the cache with new data.
+    fn write(
+        &mut self,
+        map: HashMap<i64, AuthCredentialResponse>,
+    ) -> Result<(), CredentialsCacheError>;
+}
+
+#[derive(Default)]
+pub struct InMemoryCredentialsCache {
+    map: HashMap<i64, AuthCredentialResponse>,
+}
+
+impl CredentialsCache for InMemoryCredentialsCache {
+    fn clear(&mut self) -> Result<(), CredentialsCacheError> {
+        self.map.clear();
+        Ok(())
+    }
+
+    fn get(
+        &self,
+        key: &i64,
+    ) -> Result<Option<&AuthCredentialResponse>, CredentialsCacheError> {
+        Ok(self.map.get(key))
+    }
+
+    fn write(
+        &mut self,
+        mut map: HashMap<i64, AuthCredentialResponse>,
+    ) -> Result<(), CredentialsCacheError> {
+        std::mem::swap(&mut self.map, &mut map);
```

Comment:
Fancy pants. What's wrong with self.map = map ? :'-)
```suggestion
        self.map = map;
```

(this will make it unnesessary for map to be mut too)